File: practice/algorithms/dynamic_programming/sherlock_and_cost/sherlock_and_cost.cpp
#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>
#include <algorithm>

using namespace std;

int main()
{
    int t, n;
    cin >> t;
    
    while (t--)
    {
        cin >> n;
        vector<int> b(n);
        for (int pos = 0; pos < n; ++pos)
        {
            cin >> b[pos];
        }
        
        int s1 = 0, s2 = 0;
        vector<int> b1(n);
        vector<int> b2(n);
        
        for (int pos = 1; pos < n; ++pos)
        {
            s1 = max(abs(1 - b[pos - 1]) + b2[pos - 1], b1[pos - 1]);
            s2 = max(abs(b[pos] - 1) + b1[pos - 1], abs(b[pos] - b[pos - 1]) + b2[pos - 1]);
            b1[pos] = s1;
            b2[pos] = s2;
        }
        
        cout << max(s1, s2) << endl;
    }
    return 0;
}

File: practice/algorithms/dynamic_programming/sherlock_and_cost/sherlock_and_cost.py
#!/usr/bin/env python

t = int(raw_input().strip())
for _ in range(t):
    n = int(raw_input().strip())
    b = map(int, raw_input().strip().split())
    s1 = 0
    s2 = 0
    b1 = [0 for _ in range(n)]
    b2 = [0 for _ in range(n)]
    for pos in range(1, n):
        s1 = max(abs(1 - b[pos - 1]) + b2[pos - 1], b1[pos - 1])
        s2 = max(abs(b[pos] - 1) + b1[pos - 1], abs(b[pos] - b[pos - 1]) + b2[pos - 1])
        b1[pos] = s1
        b2[pos] = s2
    print max(s1, s2)

